pkg/taskservice/mysql_task_storage.go

```diff
@@ -0,0 +1,640 @@
+		return 0, nil
+	}
+
+	tx, err := m.db.Begin()
+	if err != nil {
+		return 0, err
+	}
+	defer func(tx *sql.Tx) {
+		_ = tx.Rollback()
+	}(tx)
+	stmt, err := tx.Prepare(insertAsyncTask + "(?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)")
+	if err != nil {
+		return 0, err
+	}
+
+	j, err := json.Marshal(t.Metadata.Options)
+	if err != nil {
+		return 0, err
+	}
+	update, err := stmt.Exec(
+		t.Metadata.ID,
+		t.Metadata.Executor,
+		t.Metadata.Context,
+		string(j),
+		t.ParentTaskID,
+		t.Status,
+		t.TaskRunner,
+		t.Epoch,
+		t.LastHeartbeat,
+		t.CreateAt,
+		t.CompletedAt)
+	if err != nil {
+		return 0, err
+	}
+	exec, err := tx.Exec(updateCronTask,
+		cronTask.Metadata.Executor,
+		cronTask.Metadata.Context,
+		string(j),
+		cronTask.CronExpr,
+		cronTask.NextTime,
+		cronTask.TriggerTimes,
+		cronTask.CreateAt,
+		cronTask.UpdateAt,
+		cronTask.ID)
+	if err != nil {
+		return 0, err
+	}
+	if err := tx.Commit(); err != nil {
+		return 0, err
+	}
+	affected1, err := exec.RowsAffected()
+	if err != nil {
+		panic(err)
+	}
+	affected2, err := update.RowsAffected()
+	if err != nil {
+		return 0, err
+	}
+
+	return int(affected2) + int(affected1), nil
+}
+
+func (m *mysqlTaskStorage) taskExists(taskMetadataID string) bool {
+	var exists bool
+	err := m.db.QueryRow("select exists(select * from sys_async_task where task_metadata_id=?)", taskMetadataID).Scan(&exists)
+	if err != nil {
+		panic(err)
+	}
+	return exists
+}
+
+func (m *mysqlTaskStorage) getTriggerTimes(taskMetadataID string) (uint64, error) {
+	var triggerTimes uint64
+	err := m.db.QueryRow("select trigger_times from sys_cron_task where task_metadata_id=?", taskMetadataID).Scan(&triggerTimes)
+	if err != nil {
+		if err == sql.ErrNoRows {
+			return 0, err
+		}
+		panic(err)
+	}
+	return triggerTimes, nil
+}
+
+func buildWhereClause(c conditions) string {
+	var clause string
+
+	if c.hasTaskIDCond {
+		clause = fmt.Sprintf("task_id%s%d", OpName[c.taskIDOp], c.taskID)
+	}
+
+	if c.hasTaskRunnerCond {
+		if clause != "" {
+			clause += " AND "
+		}
+		clause += fmt.Sprintf("task_runner%s'%s'", OpName[c.taskRunnerOp], c.taskRunner)
+	}
+
+	if c.hasTaskStatusCond {
+		if clause != "" {
+			clause += " AND "
+		}
+		clause += fmt.Sprintf("task_status%s%d", OpName[c.taskStatusOp], c.taskStatus)
+	}
+
+	if c.hasTaskEpochCond {
+		if clause != "" {
+			clause += " AND "
+		}
+		clause += fmt.Sprintf("task_epoch%s%d", OpName[c.taskEpochOp], c.taskEpoch)
+	}
+
+	if c.hasTaskParentIDCond {
+		if clause != "" {
+			clause += " AND "
+		}
+		clause += fmt.Sprintf("task_parent_id%s'%s'", OpName[c.taskParentTaskIDOp], c.taskParentTaskID)
+	}
+
+	return clause
+}
+
+func buildLimitClause(c conditions) string {
+	if c.limit != 0 {
+		return fmt.Sprintf(" limit %d", c.limit)
+	}
+	return ""
+}
+
+func removeDuplicateTasks(err error, tasks []task.Task) ([]task.Task, error) {
+	me, ok := err.(*mysql.MySQLError)
+	if !ok {
+		return nil, err
+	}
+	if me.Number != 1062 {
```

Comment:
What is 1062?

Reply:
It's mysql's error code. Replace it with moerr.ER_DUP_ENTRY.